modules/storages/app/services/storages/file_link_sync_service.rb

```diff
@@ -26,205 +26,89 @@
 
   def call(file_links)
-    @file_links = file_links
-    storage_file_link_hash = @file_links.group_by(&:storage_id)
-    storage_file_link_hash.each do |storage_id, storage_file_links|
-      # Add new types of storages here. We currently only support Nextcloud
-      sync_nextcloud(storage_id, storage_file_links)
+    file_links
+      .group_by(&:storage_id)
+      .map { |storage_id, storage_file_links| sync_nextcloud(storage_id, storage_file_links) }
+      .reduce(ServiceResult.success(result: [])) do |state, sync_result|
+      state >> ->(state_result) do
+        sync_result.match(
+          on_success: ->(sr) { ServiceResult.success(result: state_result + sr) },
+          on_failure: ->(_) { ServiceResult.success(result: state_result) }
+        )
+      end
     end
-
-    @service_result
   end
 
   private
 
-  # Get the OAuthClientToken that will authenticate us against Nextcloud
-  def get_connection_manager(storage_id)
-    oauth_client = OAuthClient.find_by(integration_id: storage_id)
-    ::OAuthClients::ConnectionManager.new(user: @user, oauth_client:)
+  def sync_nextcloud(storage_id, file_links)
+    ::Storages::Peripherals::StorageRequests
+      .new(storage: ::Storages::Storage.find(storage_id))
+      .files_info_query
+      .call(user: @user, file_ids: file_links.map(&:origin_id))
+      .map(&to_hash)
+      .match(
+        on_success: set_file_link_permissions(file_links),
+        on_failure: ->(_) {
+          ServiceResult.success(result: file_links.map do |file_link|
+            file_link.origin_permission = :error
+            file_link
+          end)
+        }
+      )
   end
 
-  def get_oauth_client_token(connection_manager)
-    access_token_service_result = connection_manager.get_access_token # No scope for Nextcloud...
-    if access_token_service_result.failure?
-      return nil
-    end
-
-    access_token_service_result.result
+  def to_hash
+    ->(file_infos) { file_infos.index_by { |file_info| file_info.id.to_s }.to_h }
   end
 
-  # Sync a Nextcloud storage
-  def sync_nextcloud(storage_id, file_links)
-    connection_manager = get_connection_manager(storage_id)
-    oauth_client_token = get_oauth_client_token(connection_manager)
-    if oauth_client_token.nil?
-      set_error_for_file_links(file_links)
-      return
-    end
-
-    nextcloud_request_result = connection_manager.request_with_token_refresh(oauth_client_token) do |token|
-      response = request_files_info(token, file_links.map(&:origin_id))
-      Rails.logger.debug { "Got nextcloud filesinfo response: #{response.inspect}" }
-      # Parse HTTP response an return a ServiceResult with:
-      #   success: result= Hash with Nextcloud filesinfo (name of endpoint) data
-      #   failure: result= :error or :not_authorized
-      parse_files_info_response(response)
-    end
-    # Pass errors from Nextcloud into service result
-    @service_result.merge!(nextcloud_request_result)
+  def set_file_link_permissions(file_links)
```

Comment:
🟡 In general, having a lambda defined in a method, means that every time the method is called a new lambda is created. Again, this can put pressure on the GarbageCollector and generate a large number of trash objects.

One way around this is to define them as constants and refer to those.

Reply:
Could you give an example please?

Reply:
For the file info one, this gets tricky as it depends on instance methods, so the natural solution would be not making it a lambda. One workaround would be to make the method just return the lambda, assign its return to a local variable and reuse that variable.

```ruby
def set_link_permissions
  ->(file_links) { magic_with(file_links) }
end

def call
  permissions_setter = set_link_permissions
  Request.match(on_success: permissions_setter.call(file_links), on_failure: panic! }
end

Reply:
Okay, understood. In this case the function is only called as many times as there are storages configured in the specific project. I would think, that it is therefor okay for now. wdyt? It's hard to extract right now, because it depends on the list of file links for the selected storage which are kind of partially applied. The only place would be inside the `sync_nextcloud` function, but it would be only called once.

Reply:
I'm all for going forward as it is and later on addressing it. =)